### PR TITLE
add value_balance to builder

### PR DESCRIPTION
**Pull request by AloeareV**

### CODE REVIEW

File: src/builder.rs
Comment (daira) [inline src/builder.rs:224]: @nuttycom and I: Why cache this and update it in `add_spend`/`add_recipient`, rather than recomputing it in the `value_balance` method?

File: src/builder.rs
Comment (str4d) [inline src/builder.rs:307]: What is the use case for this method (calculating the value balance from the in-progress builder state)? It would be useful to describe intended use cases in the documentation.

File: src/builder.rs
Comment (nuttycom) [inline src/builder.rs:313]: I would prefer that this return a `ValueSum` value, which wraps an i128 internally.

**Review by nuttycom — CHANGES_REQUESTED** [review]
i64 is not sufficiently type-safe, and we already have `ValueSum` to represent the values of these kinds of balances.

**Comment by str4d** [comment]
> we already have `ValueSum` to represent the values of these kinds of balances.

`ValueSum` is not for representing value balances; that type is only ever represented abstractly in this crate. But you do raise a good point that `i64` is the wrong type to use in this API.

**Comment by AloeareV** [comment]
Switching to ValueSum necessitated adding a `Sub` impl... If the balance type is only ever represented abstractly, then maybe we add some safety checks and then still return an i64? Having a new type that is only ever used in the return of this one specific function is probably not the right answer either, if you can't use it anywhere without first doing a type conversion.

File: src/builder.rs
Comment (str4d) [inline src/builder.rs:322]: I was in the middle of writing this comment when the PR was updated:

> This sum must not be computed with `i64` values, because a value sum within a bundle is a signed 64-bit integer, which is not the same as an `i64` (which is a signed _63-bit_ integer). The `as i64` casts can cause data loss resulting in financial errors, and overflows are not handled. This should use the existing types like the builder itself does, and its return value should be user-specified like it is for `Bundle`:

My suggestion still applies:
```suggestion
    pub fn value_balance<V: TryFrom<i64>>(&self) -> Result<V, value::OverflowError> {
        let value_balance = self
            .spends
            .iter()
            .map(|spend| spend.note.value() - NoteValue::zero())
            .chain(
                self.recipients
                    .iter()
                    .map(|recipient| NoteValue::zero() - recipient.value),
            )
            .fold(Some(ValueSum::zero()), |acc, note_value| acc? + note_value)
            .ok_or(OverflowError)?;

        i64::try_from(value_balance).and_then(|i| V::try_from(i).map_err(|_| value::OverflowError))
    }
```

File: src/value.rs
Comment (str4d) [inline src/value.rs:202]: Undo this addition.

File: CHANGELOG.md
Comment (str4d) [inline CHANGELOG.md:12]: Undo this addition (and squash the commits to clean up the branch).

**Review by str4d — CHANGES_REQUESTED** [review]

**Comment by AloeareV** [comment]
`zcash_primitives::transaction::components::amount::Amount` is probably the right type in the abstract, but `zcash_primitives` depends on orchard, so there's a circular dependency issue there

File: CHANGELOG.md
Comment (AloeareV) [reply CHANGELOG.md:12]: Yes, pushed the PR addressing @nuttycom's comments, and then saw yours.

**Comment by str4d** [comment]
@AloeareV the type of a value balance is intentionally abstract in this crate, to allow different downstream consumers to use their own existing types (`zcash_primitives::transaction::components::amount::Amount` for our crates, Zebra's own type in their case). See [the documentation for the `orchard::value` module](https://docs.rs/orchard/0.2.0/orchard/value/index.html), where we document this extensively.

**Comment by str4d** [comment]
Actually, now that I think about it, this PR should add "(and [`Builder::value_balance`] for in-progress bundles)" to the end of the third bullet-point in that module documentation for `orchard::value`, now that it is adding another way that value balances are produced.

**Review by str4d — APPROVED** [review]
utACK 8011e0d57a5d245e0a0e678590dd17fd7d992664, but I am now the author of almost all of this PR's code, so it needs another review (because I do not trust myself to not get my lefts and rights mixed up).

**Review by nuttycom — APPROVED** [review]
utACK, modulo it would be good to have a test for `value_balance`

File: src/builder.rs
Comment (sellout) [inline src/builder.rs:321]: Would it help to minimize the risk of overflow if rather than `chain`, `itertools::interleave` were used?

**Review by sellout — COMMENTED** [review]
Just a question, not a requested change.

**Review by daira — CHANGES_REQUESTED** [review]
Blocking on my review tomorrow.

File: src/builder.rs
Comment (daira) [inline src/builder.rs:309]:
```suggestion
    /// can be non-zero. Each bundle's value balance is added to the transparent
    /// transaction value pool, which [must not have a negative value]. (If it were
    /// negative, the transaction would output more value than it spends.)
    ///
    /// [must not have a negative value]: https://zips.z.cash/protocol/protocol.pdf#transactions
```

File: src/builder.rs
Comment (daira) [reply src/builder.rs:321]: No, the sum of all spent values cannot exceed MAX_MONEY, and similarly for all output values. `ValueSum` enforces this.

File: src/builder.rs
Comment (daira) [inline src/builder.rs:322]: This is correct because `i64` can represent all of $[-$ MAX_MONEY, MAX\_MONEY $]$.

**Review by daira — APPROVED** [review]
utACK; also checked `src/value.rs` for overflow potential.

File: src/builder.rs
Comment (daira) [reply src/builder.rs:322]: I would also be fine with this returning `Result<ValueSum, OverflowError>` (which we already have at line 321).

File: src/builder.rs
Comment (daira) [reply src/builder.rs:309]: I forgot to commit this before merging. See #359.

**Pull request merged**
